### ISSUE TITLE
[Bug]: Missing support for Persistent Permissions with the File System Access API as of Chrome 122

### ISSUE THREAD

**Issue by Jaifroid** [electron/electron#41957]
### Preflight Checklist

- [X] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/main/CONTRIBUTING.md) for this project.
- [X] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/electron/electron/issues) for a bug report that matches the one I want to file, without success.

### Electron Version

30.0.0

### What operating system are you using?

Windows

### Operating System Version

Windows 11 (fully updated)

### What arch are you using?

x64

### Expected Behavior

Chrome 122+ now has [persistent permissions](https://developer.chrome.com/blog/persistent-permissions-for-the-file-system-access-api) for the File System Access API. In versions of Electron containing Chrome 122+, I expect to be able to set persistent permissions, either via Chrome's UI or via a Web session broker in `main`. Alternatively, I expect persistent permissions will always be granted.

### Actual Behavior

Although there is a new broker for handling File System Access permissions (see #28422 and its resolution in PR #41419 by @codebytere), using this function does not grant persistent permissions. This results in an Electron app using this API being unable to access a file or directory from a file handle stored in a previous session _**unless the user supplies a gesture**_ (e.g. mouse click). The following error shows in console: `DOMException: Failed to execute 'requestPermission' on 'FileSystemHandle': User activation is required to request permissions.`.

The current behaviour of 30.x means that developers would be forced to use Node FS APIs, and hence complicate their codebase when packaging a PWA as an Electron app.

### Additional Information

It should be noted that in Electron versions up to 29.x, it is possible to bypass the user gesture requirements, and essentially emulate persistent permissions, with the commandline switch `enable-experimental-web-platform-features`. That switch simply granted all requested permissions.

Please see https://github.com/electron/electron/issues/28422#issuecomment-2072737532 and response below it for further details about this part of the File System Access API not yet being implemented in Electron.

**Comment by bradisbell** [comment]
@codebytere If a full permissions implementation is going to be time consuming or complicated, is there any chance at a quick fix by re-enabling the previous experimental flag for full File System API access?:

```
app.commandLine.appendSwitch('enable-experimental-web-platform-features');
```

That flag would automatically grant permission.  That would at least fix the effective regression that appeared at v30, and would give us a path forward from v29 while work on a full implementation proceeds.

**Comment by Jaifroid** [comment]
I agree with @bradisbell that what appeared as a great advance with the addition of the File System Access broker has, at least in the development scope in practice, resulted in a regression. I am now stuck on v29 and cannot track latest Electron versions in my app (Kiwix JS Electron / WikiMed / Wikivoyage) until we have some fix for this issue, or until the previous workaround is reinstated. Clearly the "right" way is to fix this issue, but it would be perfectly acceptable to use the workaround as a stopgap IMHO.

**Comment by bradisbell** [comment]
@codebytere Any update?  As of August 20, we're stuck on an EOL'ed version of Electron.  Also, could you please add the `regression` label to this issue, for visibility?

Is there no way to just get the `enable-experimental-web-platform-features` to work again?

**Comment by Jaifroid** [comment]
> Is there no way to just get the `enable-experimental-web-platform-features` to work again?

I'd settle for that! It's a shame we've actually regresed with the merging of support for the File System Access API.

**Comment by pret-a-porter** [comment]
I guess it should be marked as `regression`.

I tried to pass `experimentalFeatures` and `enableBlinkFeatures` to BrowserWindow webPreferences but it does not work
1. User clicks on button and invokes showDirectoryPicker()
2. User selects folder and nothing happens. No errors/warnings.

In the same time it works fine in Electron 29 with `enable-experimental-web-platform-features `

**Comment by ikhilko** [comment]
It may be helpful for someone. I just found this code to react on file permission-restricted event from the main process:
https://www.electronjs.org/docs/latest/api/session#event-file-system-access-restricted

**Comment by bradisbell** [comment]
Still regressed... a year and a half later... 

`file-system-access-restricted` is never fired, so it can't be used in this case.

**Comment by Jaifroid** [comment]
@codebytere Could you label this as a regression? At this stage it's not just an unfinished spec, but is deeply problematic for those of stuck on v29 of Electron because the FSA persistent permissions are broken as of Electron v30. The situation was actually better prior to #41419. ☹️

**Comment by bradisbell** [comment]
Good news!  I found a way to make this work.

When you save your directory handle in IndexedDB and load it again later in a subsequent run, you can call `requestPermission()` which seems to just grant it automatically:

https://developer.mozilla.org/en-US/docs/Web/API/FileSystemHandle/requestPermission

This is working for me on Electron v37.1.0, no flags/switches.

It feels a little suspect... `file-system-access-restricted` still isn't firing for me, so I'm not sure why permission is being allowed by default.  Can anyone else confirm this is working for them?

**Comment by Jaifroid** [comment]
@bradisbell My code has always used `requestPermission` as stated in the original post of this issue. It's a basic part of the FSA verify chain, and I'm surprised you weren't using it before. As stated in the top post, in any Electron above 29.3.1 (last one known to work with the `enable-experimental-web-platform-features` flag) `requestPermission()` throws with the exception `SecurityError: Failed to execute 'requestPermission' on 'FileSystemHandle': User activation is required to request permissions`. 

If there has been a user gesture (clicking on a file selection, or on an option that triggers the `requestPermission()`) then indeed it is immediately granted without any prompts. That has always been the case. The issue we're facing is that when a user restarts the app with persisted FS permissions in indexedDB. the app can't automatically load a file and return to its original state on close because a user gesture is required. This completely deeats the point of the persistent permissions broker that @codebytere originally coded. The permissions are not effectively persisted if they need a user gesture to reactivate. When using native JS in Chrome/Edge 122+, no such user gesture is required with persisted permissions. This is definitely a regression against the equivalent functionality, and the issue persists in Electron 37.1 and 37.2 (I just re-tested). With the current state of affairs, we have better functionality in a native PWA than in an Electron app insofar as File System Access support is concerned.

**Comment by Jaifroid** [comment]
This has now become urgent because 29.3.1 now has a security vulnerability that was patched in recently released 29.3.3, but 29.3.3 has @codebytere's incomplete persistent permissions code (that unnecessarily requires user activation).

**Comment by codebytere** [comment]
hey folks i'm actively working on this and should have something by week's end.

**Comment by Jaifroid** [comment]
@codebytere That's excellent news! Looking forward to testing!

**Comment by codebytere** [comment]
Opened up https://github.com/electron/electron/pull/48170 - feel free to take a look. This should address the core issue presented here.

**Comment by Jaifroid** [comment]
For those of us who have been waiting for a long time for this fix, I can now confirm that the [PR](https://github.com/electron/electron/pull/48170) fixes the issue if you add appropriate handlers in main.js.

I have updated a test case, building on a gist by @bradisbell, that can be found [here](https://github.com/Jaifroid/electron-flesystem-permission-bug). It now has the session handlers to run under latest electron nightly (`40.0.0-nightly.20250916` or greater). Please note that this test case simply grants permission for a previously stored `FileSystemHandle` (including directory handle) ***without doing any further security checks***.
